### PR TITLE
fix: address race condition in socks5 test

### PR DESCRIPTION
**Pull request by amircybersec**
This PR attempts to resolve issue https://github.com/Jigsaw-Code/outline-sdk/issues/276

### CODE REVIEW

File: transport/socks5/stream_dialer_test.go [hunk 1]
```diff
@@ -177,7 +177,6 @@ func TestConnectWithoutAuth(t *testing.T) {
 
 	go func() {
 		err := server.Serve(listener)
-		t.Log("server is listening...")
```
Comment (fortuna) [inline transport/socks5/stream_dialer_test.go:180]: What's the race condition? This doesn't seem like it makes any difference.

**Comment by amircybersec** [comment]
@fortuna I added sync.WaitGroup to wait for go routines to complete.

File: transport/socks5/stream_dialer_test.go [hunk 2]
```diff
@@ -167,17 +167,19 @@ func testExchange(tb testing.TB, listener *net.TCPListener, destAddr string, req
...
+	var running sync.WaitGroup
+	running.Add(1)
 	// Create a SOCKS5 server.
 	server := socks5.NewServer()
 
 	// Create SOCKS5 proxy on localhost with a random port.
 	listener, err := net.Listen("tcp", "127.0.0.1:0")
 	require.NoError(t, err)
-	defer listener.Close()
+	//defer listener.Close()
```
Comment (fortuna) [inline transport/socks5/stream_dialer_test.go:178]:
```suggestion
```

File: transport/socks5/stream_dialer_test.go [hunk 3]
```diff
@@ -167,17 +167,19 @@ func testExchange(tb testing.TB, listener *net.TCPListener, destAddr string, req
 }
 
 func TestConnectWithoutAuth(t *testing.T) {
+	var running sync.WaitGroup
```
Comment (fortuna) [inline transport/socks5/stream_dialer_test.go:170]: Move this to just before the `go func()`, so it's clearer you are adding that goroutine

File: transport/socks5/stream_dialer_test.go [hunk 4]
```diff
@@ -192,9 +194,15 @@ func TestConnectWithoutAuth(t *testing.T) {
 
 	_, err = client.DialStream(context.Background(), address)
 	require.NoError(t, err)
+	// Forcefully close the listener to stop the server
+	// Otherwise defer listener.Close() will never be called
+	listener.Close()
```
Comment (fortuna) [inline transport/socks5/stream_dialer_test.go:199]: I believe this won't close if the test fails earlier. Instead, defer a function that does both the close and wait.

File: transport/socks5/stream_dialer_test.go [hunk 5]
```diff
@@ -192,9 +194,15 @@ func TestConnectWithoutAuth(t *testing.T) {
...
+	running.Wait()
 }
 
 func TestConnectWithAuth(t *testing.T) {
+	var running sync.WaitGroup
```
Comment (fortuna) [inline transport/socks5/stream_dialer_test.go:204]: Same comments

**Review by fortuna — CHANGES_REQUESTED** [review]

**Comment by amircybersec** [comment]
@fortuna thanks for the feedback. I made the changes accordingly.

**Review by fortuna — APPROVED** [review]

**Comment by fortuna** [comment]
Thanks for the fix!

**Pull request merged**